### ISSUE TITLE
Battery and memory hog

### ISSUE THREAD

**Issue by shuhaowu** [emesene/emesene#701]
Emesene appears to use a lot of memory and battery (as shown by powertop)

I use about 90 ~ 180MB of memory with emesene depending on the number of conversation open

Powertop shows that emesene is usually the second most for CPU wakeups, averaging at around 100 wake ups from idle per second, it's usually right below SignalSender, which is chrome's wakeup issue (http://code.google.com/p/chromium/issues/detail?id=77625)

**Comment by Mystro256** [comment]
The CPU issue is quite known, See #699. I'm trying to track it down, if you want to help, you can try doing a bisect, because the CPU issue is definitely a regression.

**Comment by marianoguerra** [comment]
On Sat, Jul 2, 2011 at 9:01 PM, ultimatebuster
reply@reply.github.com
wrote:

> Emesene appears to use a lot of memory and battery (as shown by powertop)
> 
> I use about 90 ~ 180MB of memory with emesene depending on the number of conversation open
> 
> Powertop shows that emesene is usually the second most for CPU wakeups, averaging at around 100 wake ups from idle per second, it's usually right below SignalSender, which is chrome's wakeup issue (http://code.google.com/p/chromium/issues/detail?id=77625)

we have to fix this:

[mariano@ganesha emesene]$ grep -R timeout_add( *
emesene.py:        self.timeout_id = glib.timeout_add(500,
emesene.py:        glib.timeout_add(500, self.session.logger.check)
gui/gtkui/AvatarManager.py:
gobject.timeout_add(time_between_frames,
gui/gtkui/Avatar.py:            self.anim_source =
gobject.timeout_add(iteran.get_delay_time(), self._advance, iteran)
gui/gtkui/Avatar.py:        self.anim_source =
gobject.timeout_add(iteran.get_delay_time(), self._advance, iteran)
gui/gtkui/Tooltips.py:            self.tag =
timeout_add(Tooltips.DELAY, self.show_tooltip, \
gui/gtkui/Tooltips.py:            self.tag =
timeout_add(Tooltips.DELAY, self.show_tooltip, \
gui/gtkui/Tooltips.py:            self.tag =
timeout_add(Tooltips.DELAY, self.show_tooltip, \
gui/gtkui/Renderers.py:
gobject.timeout_add(image.get_data('iter').get_delay_time(),
gui/gtkui/Renderers.py:
gobject.timeout_add(self._image.get_data('iter').get_delay_time(),
gui/gtkui/FileTransferBarWidget.py:        gobject.timeout_add(5,
self._move_to_left)
gui/gtkui/FileTransferBarWidget.py:        gobject.timeout_add(5,
self._move_to_right)
gui/gtkui/TextBox.py:        self.parse_timeout =
gobject.timeout_add(500, self.parse_emotes)
gui/gtkui/FileTransferWidget.py:        gobject.timeout_add(DELAY,
self.show_tooltip, \
gui/gtkui/Dialog.py:            self.tag = gobject.timeout_add(500, self.hide)
test/e3_example.py:        gobject.timeout_add(500,
self.session.signals._handle_events)

we use timeout_add a lot

**Comment by Sbte** [comment]
Actually, there's even more if you look for just timeout_add, since there's also timeout_add_seconds

**Comment by marianoguerra** [comment]
On Sat, Jul 2, 2011 at 11:19 PM, Sbte
reply@reply.github.com
wrote:

> Actually, there's even more if you look for just timeout_add, since there's also timeout_add_seconds

yes, but waking up every second or more is not that bad if you really
have to do it.

**Comment by c10ud** [comment]
i started the add_seconds campaign some time ago, but not every timeout can be converted without issues, because the resolution really is different.

as for being a "memory and cpu" hog, you should check if your contact list is full of animated avatars and such things, that's the price to pay. also: 90mb memory hog? considering there's python, a whole bunch of libs and webkit...to me it's no hog at all

**Comment by Sbte** [comment]
Just tried something:

Opening a conversation increases memory usage by 10+ mb, but closing it doesn't decrease it. If you open a new conversation window, it again increases, but now by 0.5 mb. It doesn't decrease when you close it. So at least the conversation window doesn't get destroyed properly. Might be the same story with other parts of emesene.

This is also what me and arielj saw when we found out that signals weren't unsubscribed properly. Problem then was the we didn't get the method called on dead object warning either, so the object wasn't actually dead (AdiumTextbox). So why doesn't it get destroyed properly, and how can we fix this?

**Comment by shuhaowu** [comment]
180 is an issue, though. No animated avatars.
On 2011-07-03 4:28 AM, "c10ud" <
reply@reply.github.com>
wrote:

**Comment by c10ud** [comment]
sbte, first conv loads webkit, and that's a huge thing. then there's
so much garbage collection... difficult to track, and dead stuff is
really not so easy to reach unless you want to do really hard debug

2011/7/3 ultimatebuster
reply@reply.github.com:

**Comment by Sbte** [comment]
True, I agree on the first time load. Nothing to do about that. 0.5 MB per conversation is something we could do something about though. Is there something equivalent to the C++ delete in python? Or maybe we could implement something similar to that in our extension thingy?

Ideally, there should be no memory leaks that cause the need of 5x the memory of what it needs at startup. Which is what it usually needs at the end of the day for me if I talk to a few people and close conversations regularly.

Are there tools to detect this? And suggested ways to handle it? I thought that as as soon as an object went out of scope, it was automatically destroyed. Does the extension system prevent this? Or do we do something wrong?

**Comment by c10ud** [comment]
surely something is wrong, like some references being kept on some
object, as for the tools, i wouldn't know, pythonically speaking..

2011/7/3 Sbte reply@reply.github.com:

**Comment by Mystro256** [comment]
Just to touch back on CPU issues, I find it odd receiving offline messages take much more CPU power than normal messages. Is this due to Papyon? Can someone else confirm this?

@Sbte
Well Python isn't perfect, so it can collect garbage through a few "loopholes"

Anyway, does this help at all? http://www.lshift.net/blog/2008/11/14/tracing-python-memory-leaks

**Comment by shuhaowu** [comment]
A suggestion, maybe we could move some code to a c/c++ extension, or even
just cython ?
On 2011-07-03 5:26 PM, "Mystro256" <
reply@reply.github.com>
wrote:

**Comment by marianoguerra** [comment]
On Sun, Jul 3, 2011 at 10:58 PM, ultimatebuster
reply@reply.github.com
wrote:

> A suggestion, maybe we could move some code to a c/c++ extension, or even
> just cython ?

that doesn't fix anything, memory leaks happen in C++ too

about tools to work with memory leaks mainly:
http://docs.python.org/library/gc.html

about what to do, just look at the code and make sure we don't keep
references in one object to other object when we may need it only once
or maybe in a callback, if you need it in a callback then passing it
when generating the event is the best thing.

also using weakrefs where it makes sense.

also start looking for places to subscribe to events generated while
closing stuff (like "on conversation close") and do the proper cleanup

about why offline messages take more cpu, it's because it requires
call to a web services and lot of xml parsing.

my guess is that we are keeping some references somewhere to some
objects for each conversation and that is making the memory grow.

**Comment by c10ud** [comment]
how about the LOLcode port? do we still have a branch?

2011/7/4 marianoguerra
reply@reply.github.com:

**Comment by c10ud** [comment]
2.11.7 addresses these issues, we cleanup lots of stuff now, but we're unable to go further, hence i close this issue
